tests/mgbench/benchmark.py

```diff
@@ -175,136 +536,268 @@ def filter_benchmarks(generators, patterns):
+                    args.num_workers_for_benchmark,
+                    "concurrent clients.",
                 )
-                count = int(cached_count["count"] * args.single_threaded_runtime_sec / cached_count["duration"])
+                vendor.start_benchmark(dataset.NAME + dataset.get_variant() + "_" + workload.name + "_" + query)
+                if args.warmup_run:
+                    warmup(client)
+                ret = client.execute(
+                    queries=get_queries(func, count),
+                    num_workers=args.num_workers_for_benchmark,
+                )[0]
+                usage = vendor.stop(dataset.NAME + dataset.get_variant() + "_" + workload.name + "_" + query)
+                ret["database"] = usage
+                ret["query_statistics"] = query_statistics
 
-            # Benchmark run.
-            print("Sample query:", get_queries(func, 1)[0][0])
-            print(
-                "Executing benchmark with",
-                count,
-                "queries that should yield a single-threaded runtime of",
-                args.single_threaded_runtime_sec,
-                "seconds.",
+                # Output summary.
+                print()
+                print("Executed", ret["count"], "queries in", ret["duration"], "seconds.")
+                print("Queries have been retried", ret["retries"], "times.")
+                print("Database used {:.3f} seconds of CPU time.".format(usage["cpu"]))
+                print("Database peaked at {:.3f} MiB of memory.".format(usage["memory"] / 1024.0 / 1024.0))
+                print("{:<31} {:>20} {:>20} {:>20}".format("Metadata:", "min", "avg", "max"))
+                metadata = ret["metadata"]
+                for key in sorted(metadata.keys()):
+                    print(
+                        "{name:>30}: {minimum:>20.06f} {average:>20.06f} "
+                        "{maximum:>20.06f}".format(name=key, **metadata[key])
+                    )
+                log.success("Throughput: {:02f} QPS".format(ret["throughput"]))
+
+                # Save results.
+                results_key = [
+                    dataset.NAME,
+                    dataset.get_variant(),
+                    group,
+                    query,
+                    WITHOUT_FINE_GRAINED_AUTHORIZATION,
+                ]
+                results.set_value(*results_key, value=ret)
+
+        ## If there is need for authorization testing.
+        if args.no_authorization:
+            print("Running query with authorization")
+            vendor.start_benchmark("authorization")
+            client.execute(
+                queries=[
+                    ("CREATE USER user IDENTIFIED BY 'test';", {}),
+                    ("GRANT ALL PRIVILEGES TO user;", {}),
+                    ("GRANT CREATE_DELETE ON EDGE_TYPES * TO user;", {}),
+                    ("GRANT CREATE_DELETE ON LABELS * TO user;", {}),
+                ]
+            )
```

Comment:
Duplicated code, when the runner is improved, this should not be repeated, in this case, very good place to be DRY